src/android/Geolocation.java

```diff
@@ -46,46 +48,101 @@ public class Geolocation extends CordovaPlugin implements OnLocationResultEventL
+        if(checkGMS()) {
+            fusedLocationClientGms = LocationServices.getFusedLocationProviderClient(cordova.getActivity());
+        } else if (checkHMS()) {
+            fusedLocationProviderClientHms = com.huawei.hms.location.LocationServices.getFusedLocationProviderClient(cordova.getActivity());
+        }
     }
 
     @Override
     public boolean execute(String action, JSONArray args, CallbackContext callbackContext) throws JSONException {
-        if(!checkGooglePlayServicesAvailable(callbackContext)) {
-            return false;
-        }
+        this.cordova.getActivity().runOnUiThread(() -> {
+            try {
+                if ("getLocation".equals(action)) {
+                    flowGetLocationGmsOrHms(args, callbackContext);
+                } else if ("addWatch".equals(action)) {
+                    flowAddWatchGmsOrHms(args, callbackContext);
+                } else if ("clearWatch".equals(action)) {
+                    flowClearWatchGmsOrHms(args, callbackContext);
+                }
+            } catch (Exception ex) {
+                callbackContext.error(ex.getMessage());
+            }
+        });
 
-        if ("getLocation".equals(action)) {
-            int id = args.getString(3).hashCode();
-            LocationContext lc = new LocationContext(id, LocationContext.Type.RETRIEVAL, args, callbackContext, this);
-            locationContexts.put(id, lc);
+        return true;
+    }
 
+    /**
+     * Decide if getLocation will be executed from GMS services (Google Services) or HMS (Huawei Services)
+     *
+     * @param args arguments received from execute
+     * @param callbackContext the callbackContext
+     */
+    private void flowGetLocationGmsOrHms(JSONArray args, CallbackContext callbackContext) throws JSONException {
+        int id = args.getString(3).hashCode();
+
+        if (checkGMS()) {
+            LocationContext lc = new LocationContext(id, LocationContext.Type.RETRIEVAL, args, callbackContext, this, LocationContext.DeviceType.GOOGLE_SERVICES);
+            locationContexts.put(id, lc);
             if (hasPermission()) {
-                getLocation(lc);
+                getLocationGms(lc);
             } else {
                 PermissionHelper.requestPermissions(this, id, permissions);
             }
-
-        } else if ("addWatch".equals(action)) {
-            int id = args.getString(0).hashCode();
-            LocationContext lc = new LocationContext(id, LocationContext.Type.UPDATE, args, callbackContext, this);
+        } else if (checkHMS()) {
+            LocationContext lc = new LocationContext(id, LocationContext.Type.RETRIEVAL, args, callbackContext, this, LocationContext.DeviceType.HUAWEI_SERVICES);
             locationContexts.put(id, lc);
+            if (hasPermission()) {
+                getLocationHms(lc);
+            } else {
+                PermissionHelper.requestPermissions(this, id, permissions);
+            }
+        }
```

Comment:
Before this change, we checked for Google Services, and if the device didn't have them, we showed a pop-up to the end-user telling them they needed to install or update the Google Services on their device.

What will happen now if the device doesn't either one (Google Services or Huawei Services)?

Reply:
You are right I need to add a validation, if doesn't have google either Huawei we should display the pop-up as before.